### PR TITLE
Int narrow

### PR DESCRIPTION
**Pull request by akallabeth**

### CODE REVIEW

File: libfreerdp/common/settings.c [hunk 1]
```diff
@@ -1211,26 +1211,30 @@ BOOL freerdp_settings_set_value_for_name(rdpSettings* settings, const char* name
...
+			if (!freerdp_settings_set_uint16(settings, (FreeRDP_Settings_Keys_UInt16)index,
+			                                 (UINT16)uval))
 				return parsing_fail(name, "UINT16", value);
 			return TRUE;
 
 		case RDP_SETTINGS_TYPE_INT16:
 			if (!value_to_int(value, &ival, INT16_MIN, INT16_MAX))
 				return parsing_fail(name, "INT16", value);
-			if (!freerdp_settings_set_int16(settings, (FreeRDP_Settings_Keys_Int16)index, ival))
+			if (!freerdp_settings_set_int16(settings, (FreeRDP_Settings_Keys_Int16)index,
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/common/settings.c:1222]: generated `enum`, but currently no values defined.
false positive.

**Comment by akallabeth** [comment]
rebased pr and silenced a tidy warning.

File: libfreerdp/crypto/x509_utils.c [hunk 2]
```diff
@@ -663,46 +659,42 @@ static BYTE* x509_utils_get_pem(const X509* xcert, const STACK_OF(X509) * chain,
...
 	{
 		WLog_ERR(TAG, "error allocating pemCert");
 		goto fail;
 	}
 
-	ERR_clear_error();
-	status = BIO_read(bio, pemCert, length);
-
-	if (status < 0)
+	while (offset < length)
```
Comment (mfleisz) [inline libfreerdp/crypto/x509_utils.c:673]: Wouldn't it be better to ditch the loop altogether and just use `BIO_pending` to query the necessary buffer size?

File: libfreerdp/crypto/certificate.c [hunk 3]
```diff
@@ -1281,7 +1297,7 @@ rdpCertificate* freerdp_certificate_new_from_x509(const X509* xcert, const STACK
 
 	if (chain)
 	{
-		cert->chain = sk_X509_dup(chain);
+		cert->chain = sk_X509_deep_copy(chain, X509_dup, X509_free);
```
Comment (mfleisz) [inline libfreerdp/crypto/certificate.c:1300]: Do we really need to make a deep copy here?
Think `x509_chain_up_ref` should work just fine

File: libfreerdp/crypto/certificate.c [hunk 4]
```diff
@@ -1172,7 +1180,15 @@ void certificate_free_int(rdpCertificate* cert)
 	if (cert->x509)
 		X509_free(cert->x509);
 	if (cert->chain)
+	{
+		const int count = sk_X509_num(cert->chain);
```
Comment (mfleisz) [inline libfreerdp/crypto/certificate.c:1184]: All of the code inside the if could be replaced by `sk_X509_pop_free(cert->chain, x509_free);`

File: libfreerdp/crypto/x509_utils.c (hunk 2 above)
Comment (akallabeth) [reply libfreerdp/crypto/x509_utils.c:673]: actually, on a better look, the whole function is unused :grin:

**Comment by akallabeth** [comment]
fixed issues pointed out by @mfleisz , rebased to current master.

**Pull request merged**